### PR TITLE
KAFKA-8392: Fix old metrics leakage by brokers that have no leadership over any partition for a topic

### PR DESCRIPTION
**Pull request by tuvtran**
Addresses: https://issues.apache.org/jira/browse/KAFKA-8392

- Added `removeOldLeaderMetrics` in `BrokerTopicStats` to remove `MessagesInPerSec`, `BytesInPerSec`, `BytesOutPerSec` for any broker that is no longer a leader of any partition for a particular topic
- Modified `ReplicaManager` to remove the metrics of any topic that the current broker has no leadership (meaning the broker either becomes a follower for all of the partitions in that topic or stops being a replica)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala:150]: The second type can be `Metric` to be generalized.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: There's still a potential race condition between line 173 and 174: a thread can kicks in between and remove the metric.

I think we can just do:

```
val metric: Metric = metricTypeMap.remove(metricType)
if (metric != null) {
   removeMetric(metricTypeMap, ...);
}
```

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: Check against `partitionsBecomeLeader` would not be sufficient: there maybe some existing partitions whose leader does not change, whereas `partitionsBecomeLeader` only contains the partitions that leader would be migrated in.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: We can use some scala tricks here, rather than using a mutable.Set.

File: core/src/test/scala/unit/kafka/metrics/MetricsTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/metrics/MetricsTest.scala]: Are these intentional?

**Review by guozhangwang — COMMENTED** [review]
For the unit test failure, I'd suggest we split:

```
    TestUtils.consumeTopicRecords(servers, topic, nMessages * 2)
```

Into two calls and each with `nMessages`, one before line 113 so that the metrics can be created.

**Comment by guozhangwang** [comment]
cc @ijuma for another look.

File: core/src/test/scala/unit/kafka/metrics/MetricsTest.scala
Comment (tuvtran) [reply core/src/test/scala/unit/kafka/metrics/MetricsTest.scala]: Previously this particular line caused the test to fail because `meterCount` doesn't work if `bytesOut` are not initialized

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Would you think that `leaderPartitionsIterator` is more sufficient in this case?

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Noted!

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Noted

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:150]: I tried the `Metric` type but the line
```
topicStats(topic).bytesOutRate.mark(value)
```
caused a build failure because the compiler doesn't recognize `mark` is member of the `Metric` type

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (guozhangwang) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Yes, that should be sufficient. Note that this is defined as a function so each time you call it, this iterator is dynamically created from `nonOfflinePartitionsIterator` which is in turn created from `allPartitions`.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (guozhangwang) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala:150]: Ack, let's use `Meter` then.

**Comment by guozhangwang** [comment]
Some meta comments:

1. I found `bytesRejectedRate / failedProduceRequestRate / totalProduceRequestRate` is also only marked for leader appending messages, should we do the same for it then?

2. I'm also wondering if we should generalize to create all per-topic metrics `lazily` and remove them `eagerly`, e.g. if the broker no longer hose any follower replicas of a certain topic, then we should also remove the corresponding replicationBytesInRate / OutRate as well. 

cc @junrao

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: nit: empty line.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: This function seems not used?

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: nit: name `topicNotToInclude` is a bit confusing, how about just `leaderTopicSet`?

File: core/src/test/scala/unit/kafka/metrics/MetricsTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/metrics/MetricsTest.scala:21]: This change seems not necessary?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: `partition1Replicas = Seq[Integer](1, 0)` since upon starting up the first replica in the list would be the leader.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (guozhangwang) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Could we add a unit test for handling stop-replica request as well?

**Review by guozhangwang — COMMENTED** [review]
@tuvtran Thanks for the updated PR! I've made another pass on it and it looks very good now. Left a few more comments.

**Comment by guozhangwang** [comment]
@tuvtran The pr LGTM. ping another pair of eyes to review before merging. cc @junrao

**Comment by tuvtran** [comment]
retest this please

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (junrao) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Now that we have KIP-392, it's possible for the follower to serve the fetch request. So, we probably want to keep this metric.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (junrao) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: (1) If a broker is only the follower for a topic, we will hit this code before the corresponding topic stat is created (this is lazily created when records are written and read). So, in that case, the topic stat is not present at this present. Later on, the topic stat will be added and there is no logic to remove the unneeded metrics. 

(2) We probably want to convert partitionsBecomeFollower.map(_.topic) to a set before calling diff.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (junrao) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: Similarly, if a broker is not the leader for any partition in a topic, there is no need to keep the ProduceMessageConversionsPerSec and ReplicationBytesOutPerSec metric. If a broker is not the follower for any partition, there is no need to keep the ReplicationBytesInPerSec metric.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (junrao) [inline core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: To be defensive, we want to check topicMetrics is not null before using it.

**Review by junrao — COMMENTED** [review]
@tuvtran : Thanks for the PR. Left a few comments below.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Thanks for the comment! I've returned those 2 metrics to the previous state.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Thanks! I've added a new method to remove metrics not needed for a leader-only broker: https://github.com/apache/kafka/pull/6977/files#diff-449a86e6eeefa27f776bb8dab30f6b2eR247

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Thanks! I've added the check.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (tuvtran) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: (1) If a broker is only the follower for a topic, then `bytesOutRate` and `replicationBytesOutRate` will not be created until it becomes a leader for some topics. If my assumption is right, don't we only need to remove the unneeded metrics when it stops becoming the leader and the logic for that will be executed when there is a `leaderAndIsrRequest`?

(2) Added the fix. Thank you!

Comment (guozhangwang) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: We've changed `messagesInRate` / `bytesInRate` inside the topic stat to be also lazily created: i.e. even if the topic stat object is created, these three metrics would not be created until they are being recorded. So if a broker is only follower of a topic, it would not create the above three metrics.

File: core/src/main/scala/kafka/server/KafkaRequestHandler.scala
Comment (guozhangwang) [reply core/src/main/scala/kafka/server/KafkaRequestHandler.scala]: Hmm.. why do we change the `messagesInRate` and `bytesInRate` here? @junrao was referring to the fetch requests.

In fact, I think it is still okay to make these metrics as lazily created within the topic stat even with KIP-392. What @junrao meant is that in `removeOldLeaderMetrics` we only remove the first two.

Generally speaking, I think @junrao 's other comment (https://github.com/apache/kafka/pull/6977#discussion_r302775488) was referring to:

1. MessagesInPerSec / BytesInPerSec and BytesRejectedPerSec (this is for append-only) are only used for leader. And similarly FailedProduceRequestRate, TotalProduceRequestRate, ProduceMessageConversionsPerSec and ReplicationBytesOutPerSec is only used for leader.

2. ReplicationBytesInPerSec is only used for follower.

3. BytesOutRate and FetchMessageConversionsRate are used for both leader / follower. And similarly FailedFetchRequestRate and TotalFetchRequestRate are used for both leader / follower.

We can, make all of those metrics within the topic stat to be lazily created, while in removeOldLeaderMetrics only remove group 1), and in removeOldFollowerMetrics only remove group 2), and we can potentially remove the whole topic stat (including 1,2,3) if we know this broker is not hosting leader or follower partitions of this topic.

**Comment by tuvtran** [comment]
retest this please

**Comment by tuvtran** [comment]
Hi @junrao @guozhangwang the code is ready

**Review by guozhangwang — APPROVED** [review]
LGTM! Thanks @tuvtran

**Pull request merged**
